src/raftstore/store/peer.rs

```diff
@@ -1255,15 +1266,27 @@ impl Peer {
 
         match change_type {
             ConfChangeType::AddNode => {
-                status.progress.insert(peer.get_id(), Progress::default());
+                if let Some(mut progress) = status.learner_progress.remove(&peer.get_id()) {
+                    // For promote learner to voter.
+                    progress.is_learner = false;
+                    status.progress.insert(peer.get_id(), progress);
+                } else {
+                    status.progress.insert(peer.get_id(), Progress::default());
+                }
             }
             ConfChangeType::RemoveNode => {
+                if peer.get_is_learner() {
+                    // If the node is a learner, we can return directly.
+                    return Ok(());
+                }
                 if status.progress.remove(&peer.get_id()).is_none() {
                     // It's always safe to remove a unexisting node.
                     return Ok(());
                 }
             }
-            ConfChangeType::AddLearnerNode => unimplemented!(),
+            ConfChangeType::AddLearnerNode => {
+                return Ok(());
```

Comment:
Does it mean it should not be reachable?

Reply:
No, it's reachable. `return OK` means `AddLearner` won't break the safe rule of conf change, described at line 1222.